File: ts/nestjs/src/http/problem.ts
import { STATUS_CODES } from 'node:http';
import type { Response } from 'express';
import type { FieldErrorDto } from '../dto/problem.dto.js';

export function sendProblem(
  response: Response,
  status: number,
  code: string,
  fields?: FieldErrorDto[],
  headers: Record<string, string> = {},
): void {
  if (response.headersSent) {
    response.destroy();
    return;
  }
  for (const [key, value] of Object.entries(headers)) {
    if (
      !['content-type', 'content-length', 'x-request-id'].includes(
        key.toLowerCase(),
      )
    )
      response.setHeader(key, value);
  }
  response
    .status(status)
    .type('application/problem+json')
    .json({
      type: 'about:blank',
      title: STATUS_CODES[status] ?? 'HTTP Error',
      status,
      code,
      request_id: response.getHeader('X-Request-ID'),
      ...(fields ? { errors: fields } : {}),
    });
}

File: java/spring-boot/src/main/java/com/backendtemplate/dto/Problem.java
package com.backendtemplate.dto;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.List;

@JsonInclude(JsonInclude.Include.NON_NULL)
public record Problem(String type, String title, int status, String code,
        @JsonProperty("request_id") String requestId, List<FieldError> errors) {}
